docs/05_best-practices/08_abi/01_abi-code-generator-attributes-explained.md

```diff
@@ -93,3 +93,24 @@ extern "C" {
 
 The code above will mark a function declaration as being a WebAssembly import.  This allows for other compilation modes to specify which functions are import only (i.e. do not link) without having to maintain a secondary file with duplicate declarations.
 
+## [[eosio::action, eosio::read-only]]
+The `read-only` attribute marks a method which has been defined as an action as a read-only action.
+
+Example:
+
+```cpp
+[[eosio::action, eosio::read-only]]
+std::vector<my_struct> get() {
+   std::vector<my_struct> ret;
+   // retrieve blockchain state and populate the ret vector
+   return ret; 
+}
+```
+
+Contract actions tagged read-only:
+* Cannot call insert/update (write) functions on the `Multi-Index API`, nor the `Key Value API`.
+* Cannot call `deferred transactions`.
+* Cannot call `inline actions`.
+* The returned data size is limited by the action return value sizes. By default these are set to 256 bytes by `default_max_action_return_value_size`.
+
+The `eosio-cpp` and `eosio-cc` tools will generate an error and terminate compilation if an action tagged read-only attempts to use a write API. However, if the command-line override option `--warn-action-read-only` is used, the `eosio-cpp` and `eosio-cc` tools will issue a warning and continue compilation.
```

Comment:
the error/warning is issued only for "attempts to use a write API"? or for all the other restrictions listed previously

```
Contract actions tagged read-only:

* Cannot call insert/update (write) functions on the `Multi-Index API`, nor the `Key Value API`.
* Cannot call `deferred transactions`.
* Cannot call `inline actions`.
```

Reply:
re-worded

Reply:
An error will indeed be generated for all of the above conditions.  The full list of forbidden function is:

- "set_resource_limits"
- "set_wasm_parameters_packed"
- "set_resource_limit"
- "set_proposed_producers"
- "set_proposed_producers_ex"
- "set_blockchain_parameters_packed"
- "set_parameters_packed"
- "set_kv_parameters_packed"
- "set_privileged"
- "db_store_i64"
- "db_update_i64"
- "db_remove_i64"
- "db_idx64_store"
- "db_idx64_update"
- "db_idx64_remove"
- "db_idx128_store"
- "db_idx128_update"
- "db_idx128_remove"
- "db_idx256_store"
- "db_idx256_update"
- "db_idx256_remove"
- "db_idx_double_store"
- "db_idx_double_update"
- "db_idx_double_remove"
- "db_idx_long_double_store"
- "db_idx_long_double_update"
- "db_idx_long_double_remove"
- "kv_erase"
- "kv_set"
- "send_deferred"
- "send_inline"
- "send_context_free_inline"

This includes several privileged functions not ordinarily used outside of system contracts, as well as the aforementioned `send_deferred` and `send_inline`.  Categorical mention should probably be made of all of these, and ideally the comprehensive list incorporated into reference documentation.  (This may require adding doxygen comments, which we definitely should be doing more often.)

Reply:
updated again - I agree re reference docs -> these are autogenerated from doxygen comments. the entire list is not suited to this page